test/models/test_gpt.py

```diff
@@ -135,19 +162,163 @@ def mm_decoder(
+            )
+            assert_expected(
+                layer_past_kv["v"].shape,
+                torch.Size([1, 2, 3, 2]),
+            )
+        # right shift should be switched on to prepend SOS to in_modality sequence
+        mock_right_shift.assert_called_once()
+        mock_right_shift.reset_mock()
+
+        # generate out_modality for one step
+        # take the last in_modality token as the starting token for out_modality generation
+        decode_step = 0
+        with torch.no_grad():
+            actual = gpt.fwd(
+                out_modality=in_modality[:, -1:, :],
+                out_pos_ids=torch.tensor([decode_step]).unsqueeze(0).repeat(b, 1),
+                use_cache=True,
+                causal=True,
+            )
+        # check that the key/value representation has increased by 1 unit
+        for layer_past_kv in actual.past_key_values:
+            assert_expected(
+                layer_past_kv["k"].shape,
+                torch.Size(
+                    [1, 2, 4, 2]
+                ),  # (b, n_head, in_seq_len + 1, d_model // n_head)
+            )
+            assert_expected(
+                layer_past_kv["v"].shape,
+                torch.Size([1, 2, 4, 2]),
+            )
+        # right shift should be switched off as the "SOS" token for out_modality is the last token
+        # of in_modality
+        mock_right_shift.assert_not_called()
+
+    def test_forward(
+        self,
+        gpt,
+        in_modality,
+        out_modality,
+        self_attn_mask,
+        self_head_mask,
+        logits_mask,
+    ):
+        gpt = gpt()
+
+        b, in_seq_len, _ = in_modality.shape
+        b, out_seq_len, _ = out_modality.shape
+        attn_mask = self_attn_mask(in_seq_len + out_seq_len)
+        head_mask = self_head_mask(in_seq_len + out_seq_len)
+        actual = gpt.forward(
+            in_modality=in_modality,
+            out_modality=out_modality,
+            attn_mask=attn_mask,
+            head_mask=head_mask,
+            use_cache=True,
+            causal=True,
+            right_shift=True,
+            logits_mask=logits_mask,
+        )
+        assert isinstance(actual, MultimodalGPTOutput)
+        expected = {
+            "decoder_output": {
+                "last_hidden_states": (
+                    torch.Size([1, 7, 4]),  # (b, seq_len, d_model)
+                    -0.2385,
+                ),
+                "hidden_states": None,
+                "attention_weights": None,
+                "past_key_values": (
+                    (
+                        {
+                            "k": (torch.Size([1, 2, 7, 2]), 6.0892),
+                            "v": (torch.Size([1, 2, 7, 2]), 2.6072),
+                        }
+                    ),
+                ),  # (num_layers, key/value, (b, n_head, seq_len, d_model // n_head)
+            },
+            "logits": (torch.Size([1, 10, 7]), 0.0),  # (b, tokens, seq_len)
+        }
+
```

Comment:
no assertion here?

Reply:
Added. Also added a separate test for the `logits_mask`.